mijia/src/bluetooth/mod.rs

```diff
@@ -481,30 +480,16 @@ impl BluetoothSession {
         let sensors = tree
             .into_iter()
             .filter_map(|(object_path, interfaces)| {
-                // FIXME: can we generate a strongly typed deserialiser for this,
-                // based on the introspection data?
-                let device_properties = interfaces.get("org.bluez.Device1")?;
-
-                let mac_address = device_properties
-                    .get("Address")?
-                    .as_iter()?
-                    .filter_map(|addr| addr.as_str())
-                    .next()?
-                    .to_string();
-                let name = device_properties.get("Name").map(|name| {
-                    name.as_iter()
-                        .unwrap()
-                        .filter_map(|addr| addr.as_str())
-                        .next()
-                        .unwrap()
-                        .to_string()
-                });
+                let device_properties = OrgBluezDevice1Properties::from_interfaces(&interfaces)?;
```

Comment:
(mostly rambling and probably re-hashing the discussion from your upstream PR, so feel free to ignore me)

I feel like it's a bit surprising that device_properties owns a reference into `interfaces`, but I guess it is efficient and device_properties will typically be short-lived, so it's probably fine (I also can't think of a better way to do it).

I suppose if you were only asking for a single property then you'd be using a fluent interface (`OrgBluezDevice1Properties::from_interfaces(&interfaces)?.address()?`) and it would not be surprising at all.

I suppose if OrgBluezDevice1Properties was a trait with a blanket impl over `&'a ::std::collections::HashMap<String, arg::PropMap>` then it could be written `OrgBluezDevice1Properties::address(&interfaces)`, but people would be tempted to `use OrgBluezDevice1Properties` and write `interfaces.address()?` which might be a bit fragile.

Reply:
Yeah, it looks a little weird, but seems to work well in practice so far. The other issue with the trait approach is that I want `from_interfaces` to return something that can be passed around, which is a bit of a pain if it's a trait object.

Note that `from_intefaces` is doing one level of HashMap lookup, and `address()` is doing a second level.